lib/zendesk_api/middleware/request/retry.rb

```diff
@@ -15,44 +15,74 @@ def initialize(app, options = {})
           original_env = env.dup
+          original_env[:call_attempt] ||= 0
+
           exception_happened = false
+          response = nil
+
           if @retry_on_exception
             begin
               response = @app.call(env)
-            rescue => e
+            rescue => ex
               exception_happened = true
+              exception = ex
             end
           else
+            # Allow exceptions to propagate normally when not retrying
             response = @app.call(env)
           end
 
-          if exception_happened || @error_codes.include?(response.env[:status])
+          if exception_happened
+            original_env[:call_attempt] += 1
+            seconds_left = DEFAULT_RETRY_AFTER.to_i
+            @logger&.warn "An exception happened, waiting #{seconds_left} seconds... #{exception}"
+            instrument_retry(original_env, "exception", seconds_left)
+            sleep_with_logging(seconds_left)
+            return @app.call(original_env)
+          end
 
-            if exception_happened
-              seconds_left = DEFAULT_RETRY_AFTER.to_i
-              @logger&.warn "An exception happened, waiting #{seconds_left} seconds... #{e}"
-            else
-              seconds_left = (response.env[:response_headers][:retry_after] || DEFAULT_RETRY_AFTER).to_i
-            end
+          # Loop through consecutive retryable responses (e.g., multiple 429s)
+          while response && @error_codes.include?(response.env[:status])
+            original_env[:call_attempt] += 1
+            seconds_left = (response.env[:response_headers][:retry_after] || DEFAULT_RETRY_AFTER).to_i
+            @logger&.warn "You may have been rate limited. Retrying in #{seconds_left} seconds..."
+            instrument_retry(original_env, (response.env[:status] == 429) ? "rate_limited" : "server_error", seconds_left)
+            sleep_with_logging(seconds_left)
+            response = @app.call(original_env)
+          end
 
-            @logger&.warn "You have been rate limited. Retrying in #{seconds_left} seconds..."
+          response
+        end
 
-            seconds_left.times do |i|
-              sleep 1
-              time_left = seconds_left - i
-              @logger&.warn "#{time_left}..." if time_left > 0 && time_left % 5 == 0
-            end
+        private
 
-            @logger&.warn ""
+        def instrument_retry(env, reason, delay)
+          return unless @instrumentation
+          @instrumentation.instrument(
+            "zendesk.retry",
+            {
+              attempt: env[:call_attempt],
+              endpoint: env[:url].path,
+              method: env[:method],
+              reason: reason,
+              delay: delay
+            }
+          )
+        end
 
-            @app.call(original_env)
-          else
-            response
+        def sleep_with_logging(seconds_left)
+          seconds_left.times do |i|
+            sleep 1 if seconds_left > 0
```

Comment:
We can remove this redundant conditional. If we enter inside this loop `seconds_left` times, `seconds_left > 0` must always be `true`.